nix/supervisord-cluster/topology.nix

```diff
@@ -0,0 +1,106 @@
+{ stateDir
+, graphviz
+, composition
+, localPortBase
+}:
+with composition;
+let
+  # metadata = {
+  #   inherit benchmarkingProfileName benchmarkingProfile benchmarkingTopology;
+  # };
+  producers =  cfg.producers;
+
+  topology = builtins.toFile "topology.yaml" (builtins.toJSON {
+    Producers =
+      map (n: {
+        addr = let a = n.addr or n; in if (nodes ? ${a}) then hostName a else a;
+        port = n.port or nodePort;
+        valency = n.valency or 1;
+      }) cfg.producers;
+  });
+in
+
+''
+  mkdir -p "${stateDir}/topologies"
+
+  ## 0. Generate the overall topology, in the 'cardano-ops'/nixops style:
+  #
+  args=( --topology-output "${stateDir}/topology-nixops.json"
+         --dot-output      "${stateDir}/topology.dot"
+         --size             ${toString n_hosts}
+         ${toString
+           (map (l: ''--loc ${l}'') locations)}
+       )
+  topology "''${args[@]}"
+
+  ${graphviz}/bin/neato -s120 -Tpdf \
+     "${stateDir}/topology.dot" > "${stateDir}/topology.pdf"
+
+  ## 1. Patch the nixops topology with the density information:
+  #
+  jq 'def nixops_topology_set_pool_density($topo; $density):
+        $topo *
+        { coreNodes:
+          ( .coreNodes
+          | map
+            ( . *
+              { pools:
+                (if .pools == null then 0 else
+                 if .pools == 1    then 1 else
+                    ([$density, 1] | max) end end)
+              }
+            )
+          )
+        };
+
```

Comment:
The is some Haskell code to generate the topology.
Why is the output of the Haskell tool patched with jq ?
Is it not better to do everything in Haskell ?

Reply:
The short answer is that this specific format does not exist in isolation.

The topology as generated by the haskell tool can be consumed by the AWS deployment infra, without any changes -- and that's what is valuable here.